src/backend/distributed/utils/attribute.c

```diff
@@ -0,0 +1,688 @@
+	 */
+	bool returnAllTenants = PG_GETARG_BOOL(0);
+
+	TupleDesc tupleDescriptor = NULL;
+	Tuplestorestate *tupleStore = SetupTuplestore(fcinfo, &tupleDescriptor);
+	time_t monitoringTime = time(0);
+
+	Datum values[CITUS_STATS_TENANTS_COLUMNS];
+	bool isNulls[CITUS_STATS_TENANTS_COLUMNS];
+
+	MultiTenantMonitor *monitor = GetMultiTenantMonitor();
+
+	if (monitor == NULL)
+	{
+		PG_RETURN_VOID();
+	}
+
+	LWLockAcquire(&monitor->lock, LW_EXCLUSIVE);
+
+	int numberOfRowsToReturn = 0;
+	if (returnAllTenants)
+	{
+		numberOfRowsToReturn = monitor->tenantCount;
+	}
+	else
+	{
+		numberOfRowsToReturn = Min(monitor->tenantCount, CitusStatsTenantsLimit);
+	}
+
+	for (int tenantIndex = 0; tenantIndex < monitor->tenantCount; tenantIndex++)
+	{
+		UpdatePeriodsIfNecessary(&monitor->tenants[tenantIndex], monitoringTime);
+		ReduceScoreIfNecessary(&monitor->tenants[tenantIndex], monitoringTime);
+	}
+	SafeQsort(monitor->tenants, monitor->tenantCount, sizeof(TenantStats),
+			  CompareTenantScore);
+
+	for (int i = 0; i < numberOfRowsToReturn; i++)
+	{
+		memset(values, 0, sizeof(values));
+		memset(isNulls, false, sizeof(isNulls));
+
+		TenantStats *tenantStats = &monitor->tenants[i];
+
+		values[0] = Int32GetDatum(tenantStats->colocationGroupId);
+		values[1] = PointerGetDatum(cstring_to_text(tenantStats->tenantAttribute));
+		values[2] = Int32GetDatum(tenantStats->readsInThisPeriod);
+		values[3] = Int32GetDatum(tenantStats->readsInLastPeriod);
+		values[4] = Int32GetDatum(tenantStats->readsInThisPeriod +
+								  tenantStats->writesInThisPeriod);
+		values[5] = Int32GetDatum(tenantStats->readsInLastPeriod +
+								  tenantStats->writesInLastPeriod);
+		values[6] = Int64GetDatum(tenantStats->score);
+
+		tuplestore_putvalues(tupleStore, tupleDescriptor, values, isNulls);
+	}
+
+	LWLockRelease(&monitor->lock);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * clean_citus_stats_tenants cleans the citus_stats_tenants monitor.
+ */
+Datum
+clean_citus_stats_tenants(PG_FUNCTION_ARGS)
+{
+	MultiTenantMonitor *monitor = GetMultiTenantMonitor();
+	monitor->tenantCount = 0;
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * sleep_until_next_period sleeps until the next monitoring period starts.
+ */
+Datum
+sleep_until_next_period(PG_FUNCTION_ARGS)
+{
+	struct timeval currentTime;
+	gettimeofday(&currentTime, NULL);
+
+	long int nextPeriodStart = currentTime.tv_sec -
+							   (currentTime.tv_sec % CitusStatsTenantsPeriod) +
+							   CitusStatsTenantsPeriod;
+
+	long int sleepTime = (nextPeriodStart - currentTime.tv_sec) * 1000000 -
+						 currentTime.tv_usec + 100000;
+	pg_usleep(sleepTime);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * AttributeQueryIfAnnotated checks the query annotation and if the query is annotated
+ * for the tenant statistics monitoring this function records the tenant attributes.
+ */
+void
+AttributeQueryIfAnnotated(const char *query_string, CmdType commandType)
```

Comment:
is there a way to disable the annotation injection & parsing?

Reply:
Just merged #6769 I think it does what you're asking for.